pkg/store/store.go

```diff
@@ -17,44 +17,164 @@ limitations under the License.
 package store
 
 import (
+	"fmt"
+	"strings"
 	"sync"
 
 	authorizationv1 "k8s.io/api/authorization/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	v1a1 "sigs.k8s.io/referencegrant-poc/apis/v1alpha1"
 )
 
+const ExtraPurposeKey = "purpose"
+
 // AuthorizationStore interface
 type AuthorizationStore interface {
 	CheckAuthz(sar authorizationv1.SubjectAccessReview) (bool, error)
 	// AddGrant(resource, verb string, allowed bool)
 }
 
-// in-memory GraphStore
+// currently "group/resource"
+type TargetResourceGroup string
+
+// currently "ns/name"
+// type TargetNamespacedName string
+
+// This is the "For" string
+type Purpose string
+
+// Initial version of the graph - maps "from-to-for" to a map of target("to")resource names to set of subjects
+// This ignores classNames, namespace, verbs.
+type GrantGraph map[string]map[types.NamespacedName]sets.Set[v1a1.Subject]
+
+// in-memory AuthStore
 type AuthStore struct {
-	graph map[string]map[string]bool
-	mutex sync.RWMutex
+	graph GrantGraph
+	// TODO: should we support multiple purposes for the same subject ?
```

Comment:
Yes

Reply:
I mean multiple purpose for the same subject AND the same "From" & "TO". it should be very rare I think.